File: cmd/zhiyuan-data/fenduan.go
package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"

	"github.com/sunfmin/zhiyuanwiki/internal/store"
)

// fenduanCmd 从 SQLite staging 投影某省一分一段 JSON（按省份分目录），供站点做分数↔位次换算。
// 全部省份都已入库（ADR-0014），统一从 DB 投影。
func fenduanCmd(args []string) {
	fs := flag.NewFlagSet("fenduan", flag.ExitOnError)
	out := fs.String("out", filepath.Join("src", "data"), "JSON 输出目录（其下按省份 slug 分目录）")
	dbPath := fs.String("db", filepath.Join("out", "zhiyuan.db"), "SQLite staging 库")
	provSlug := fs.String("prov", "hlj", "省份 slug：hlj / zj / js …")
	_ = fs.Parse(args)
	p := mustProv(*provSlug)
	fenduanFromDB(*dbPath, *out, p)
}

// fenduanFromDB 从 SQLite staging 投影某省一分一段 JSON（每 年×科类 一个文件）。见 ADR-0014。
func fenduanFromDB(dbPath, out string, p province) {
	db, err := store.Open(dbPath)
	if err != nil {
		fatal(err)
	}
	defer db.Close()
	yds, err := db.LoadYiFenYiDuan(p.slug, p.name)
	if err != nil {
		fatal(err)
	}
	if len(yds) == 0 {
		fatal(fmt.Errorf("DB 无 %s 一分一段——先跑 `zhiyuan-data import -prov %s`", p.name, p.slug))
	}
	dir := filepath.Join(srcDir(out, p), "yifenyiduan")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		fatal(err)
	}
	for _, y := range yds {
		slug, ok := trackSlug[y.Track]
		if !ok {
			continue
		}
		outName := fmt.Sprintf("%s-%d.json", slug, y.Year)
		writeJSON(filepath.Join(dir, outName), y)
		fmt.Printf("✓ %s · %s %d 一分一段：%d 个分数段 → %s\n",
			p.name, y.Track, y.Year, len(y.Entries), outName)
	}
}

File: src/lib/fenduan.ts
// 一分一段：分数 ↔ 位次换算。纯函数，与 Go internal/hlj 逻辑镜像，客户端用。

export interface FenduanEntry {
  score: number;
  count: number;
  cumulative: number; // 累计人数 = 位次
}

export interface YiFenYiDuan {
  province: string;
  track: string;
  year: number;
  entries: FenduanEntry[]; // 按 score 升序
}

/**
 * 分数 → 位次：取"分数 ≥ score 的最小分数段"的累计人数。
 * 正确处理顶段（X以上）、缺失分（就近向上）、高于顶段（返回顶段累计）。
 */
export function scoreToRank(t: YiFenYiDuan, score: number): number | null {
  const e = t.entries;
  if (e.length === 0) return null;
  // 二分找第一个 score' >= score
  let lo = 0;
  let hi = e.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (e[mid].score >= score) hi = mid;
    else lo = mid + 1;
  }
  if (lo === e.length) return e[e.length - 1].cumulative;
  return e[lo].cumulative;
}

/**
 * 位次 → 分数：取"累计人数 ≥ rank 的最高分数段"。
 */
export function rankToScore(t: YiFenYiDuan, rank: number): number | null {
  const e = t.entries;
  if (e.length === 0 || rank < 1) return null;
  for (let i = e.length - 1; i >= 0; i--) {
    if (e[i].cumulative >= rank) return e[i].score;
  }
  return e[0].score;
}
